File: three-number-sort/ThreeNumberSort.js
var threeNumberSort = (array, order) => {
    let left = 0;
    for(const num of order) {
        let right = array.length - 1;
        while(left < right) {
            if(array[left] === num) { left++; continue; }
            while(right > left) {
                if(array[right] === num) {
                    [array[right], array[left]] = [array[left], array[right]];
                    left++; break; 
                }
                right--;
            }
        }
    }
    return array;
}

var threeNumberSort = (array, order) => {
    let right = array.length - 1;
    for(let i = order.length - 1; i > -1; i--) {
        let left = 0;
        while(left < right) {
            if(array[right] === order[i]) { right--; continue; }
            while(left < right) {
                if(array[left] === order[i]) {
                    [array[left], array[right]] = [array[right], array[left]];
                    right--; break;
                }
                left++;
            }
        }
    }
    return array;
}
exports.threeNumberSort = threeNumberSort;

File: three-number-sort/ThreeNumberSort.java
public class ThreeNumberSort {
    public int[] threeNumberSort1(int[] array, int[] order) {
        int left = 0;
        for(int num : order) {
            int right = array.length - 1;
            while(left < right) {
                if(num == array[left]) { left++; continue; }
                while(right > left) {
                    if(array[right] == num) {
                        int temp = array[left];
                        array[left] = array[right];
                        array[right] = temp;
                        left++; break;
                    }
                    right--;
                }
            }
        }
        return array;
    }

    public int[] threeNumberSort2(int[] array, int[] order) {
        int right = array.length - 1;
        for(int i = order.length - 1; i >= 0; i--){
            int left = 0;
            while(left < right) {
                if(array[right] == order[i]) { right--; continue; }
                while(left < right) {
                    if(array[left] == order[i]) {
                        int temp = array[left];
                        array[left] = array[right];
                        array[right] = temp;
                        right--; break;
                    }
                    left++;
                }
            }
        }
        return array;
    }
}
